### DISCUSSION TITLE
Adding a functional button entity

### DISCUSSION THREAD

**Discussion by AlmostInteractive** [NoeticTools/Net2HassMqtt#2] [Ideas]
Hey @RobSmyth , I was hoping I might borrow some of your genius.  I'd like to make a ButtonEntity but I'm a little lacking in experience.  I added the entity description to the `HassTypesGeneratorDefinition.cs` file and while it shows up and the button works from HA -> App, two things are still wrong:
 - I can't activate the button from the App (eg: App -> HA)
 - Every time mqtt stops or starts it publishes `/{button_name}/config` which, for some reason, activates a button press in HA (the event description is: "{button name} changed to {date_time of event}")

Here's a link to my changes, insufficient as they are.  Do you have any thoughts about how to resolve those issues?
https://github.com/AlmostInteractive/Net2HassMqtt/compare/main...AlmostInteractive:Net2HassMqtt:buttonEntity?expand=1

**Comment by AlmostInteractive** [comment]
I just found your TODOs: button and event.  :)  I have a use case for one or both!  I'll bang my head against it for awhile but I'd sure like your input.  I'm not sure I'm smart enough to do it alone.

**Comment by RobSmyth** [comment]
Yea, I have not done the button as it is an 'event' and I have not looked at how to do that. Nice you found the `HassTypesGeneratorDefinition.cs`.

I'll have a bit of a look and get back to you.

**Comment by AlmostInteractive** [comment]
I ran some experiments, evidently buttons get "pressed" (according to HA) whenever the state changes, so if it becomes Unknown or Unavailable then that will trigger a Button Press automation.  :-/  Nothing I can do about that.

I've got an event being fired within the entity code, and it's publishing "something" to MQTT, but it's not the correct something.  (I shoehorned in the entity stuff but it might give you an idea of what I'm thinking about.)

I created an event in my app like this:
```
device.HasEvent(config => config.OnModel(this)
            .WithEventProperty(Subscribe_TestButtonPressed, Unsubscribe_TestButtonPressed)
            .WithEventTypes(["press"])
            .WithFriendlyName("Test Event")
            .WithNodeId("test_event"));

...

public event EventProperty? TestButtonPressed;

    public void Fire_TestButtonPressed() {
        TestButtonPressed?.Invoke();
    }

    public void Subscribe_TestButtonPressed(EventProperty callback) {
        TestButtonPressed += callback;
    }
    
    public void Unsubscribe_TestButtonPressed(EventProperty callback) {
        TestButtonPressed -= callback;
    }
```

Currently, the Event doesn't register in HA because the `{topic}/config` is missing the `event_types` field.  I'm trying to figure that out at the moment.

**Comment by RobSmyth** [comment]
@AlmostInteractive  - I took at your PR. Nice! It makes me very happy to see that somebody can read my code and extend it. Thank you.

I took a look at a button I have on my MQTT network. The question that we need to resolve is - how doe the client app model events?

All of the entities to date are state entities. Something reports ON and in an hour if nothing is reported it is still ON. So the code tells the MQTT broker to retain this value. This means that if a client (HASS) is restarted it gets the current state. Events, I think, should not be retained as we do not want to be told the button is pressed every time HASS is restarted.

So there is the first, not so big issue - do not retain events.

In MQTT the events are otherwise pretty well the same as the states. When I press my button it will sent an event value of "single", "double", or "hold" (that I know of). After I release it, it seems to send an event value "" (empty string). So how does the model provide this?

If we use properties, as we have to date, then the application would need to do the work of setting a string property to "single" for say 100ms and then to "". It does not feel right. Hmm. I'm wondering if a public event would be the go. That is rather than adding a button property we add a button event on the class and listed to events ... that feel right. Doable but different to all the others  :-)

So a model may look something like:

```
public class MyAppModel : ObservableObject
{
    public EventHandler<Net2HassEventEventArgs> ButtonPressedEvent;   // <<< entity will subscribe to this?

    [ObservableProperty]
    private bool _switch1 = true;    // a typical notifying state property the code uses now - no change
}
```

The bit I do not like is that is exposes a N2H data type `Net2HassEventEventArgs` in the client app (sigh).

I'm still pondering ... does this make any sense? Your thoughts?

**Reply by AlmostInteractive** [reply]
Happy to help, @RobSmyth !  :-)  I also thought about temporarily setting a property and unsetting it, as you suggested, but I think that would result in two button presses and possibly two event triggers.  The code I've got in there now is listening to an event for both Button and Event entities.  It's a little hacky but it's working, I can see the event listener responding.  At the moment it's calling `EntityBase::OnEventActivated` which publishes a command.  That may work, but two things:

- No matter what I publish on MQTT, HA doesn't respond to a button message
- Event entities aren't showing up because they need `event_types` to be set on the `{topic}/config` and they're not, even if I put them into `additional_options`.  I haven't been able to trace a path from `additional_options` to the `EventConfigMqttJson` constructor to set the `EventTypes`.  Maybe you know a way?

**Comment by RobSmyth** [comment]
Sounds like we are both on the same wavelength. I'll give your code a run and get back to you.

**Comment by AlmostInteractive** [comment]
I fixed point 2 above.  :)  It's now correctly publishing the AdditionalOptions in the entity config!  It's publishing on event detection, but I don't think it's publishing the correct information.  I can't get it to trigger anything in HA.

I'm on the other side of the world so I'll catch you when I wake up.

**Comment by RobSmyth** [comment]
What I notice is that the publish payload just has "press".  Look at the differences between battery_1_charging and test_event here seen in MQTT Explorer:

![image](https://github.com/user-attachments/assets/0ec617ea-5f42-48eb-9cb0-994991aef61d)

The configuration has:

```
  "value_template": "{{ value_json.state }}"
```

but no "state" property in the payload. The battery_1_charging looks like this:

```
{
  "data": {},
  "state": "ON"
}
```

**Comment by RobSmyth** [comment]
I do not know but I think you may want to send 2 events. "press" then "" or "None" as the press is a momentary event.

FYI - I capture button press events in Node-RED by an event:

![image](https://github.com/user-attachments/assets/6255faa3-aea4-4358-994b-8dcf37d2ec55)

**Comment by RobSmyth** [comment]
I think you want a payload like:

```
{
  "data": {},
  "action": "press"
}
```

`data` is used for attributes.

**Comment by AlmostInteractive** [comment]
I updated that branch to send that payload but it's still not registering in HA.  I've been using MQTTX to manually send events over the network as well, and nothing I send to any topic I can think of registers as "Detected an Event" in HA.  The only "events" detected are when my app starts up or shuts down (birth or lwt messages):
![image](https://github.com/user-attachments/assets/44107d3b-5308-41ca-ad0d-662d651d1b5d)

Also, I added you as a collaborate on that project in case you want to make contributions to my/your repo 😂 

I think if we can figure out what message to send, then the Button will work.  Once the Button works, getting the generic Event to work should be simple.

**Comment by AlmostInteractive** [comment]
Okay, so it seems to be working with ` hacommsclient/hacomms/test_event` `{"event_type": "press"}`.  The problem is the `value_template` in the config.  When I remove it, it works!  So we need to modify/remove the `value_template` for events (and maybe buttons, haven't experimented yet).

**Comment by AlmostInteractive** [comment]
End of day update: It works.  I've got it so the `value_template` can be overridden via `HassTypesGeneratorDefinitions`.  By setting the value template to null, the event works.

Right now it's hardcoded to send "press" but it shouldn't be too hard to make it a generic payload event with "event_type" being required.  It kinda bugs me how I shoehorned in Subscribe/Unsubscribe but I'm not really sure how to change it.

**Comment by RobSmyth** [comment]
> End of day update: It works.

Great to hear :-)

>  It kinda bugs me how I shoehorned in Subscribe/Unsubscribe but I'm not really sure how to change it.

Is this the subscribe/unsubscribe methods on the model?

**Comment by RobSmyth** [comment]
I've created a branch in your repo off your working branch.

I think the code needs refactoring to make way for easier event implementation. I'm removing base class  between attributes and entities and renaming the EntityBase class to StateEntityBase class. The intent is be able to add an EventEntityBase type so there is clear separation ... see how it goes. I may abandon this ... see how it goes  :-)

**Reply by AlmostInteractive** [reply]
Oh, I like that idea!  I should have thought of that!  Yes, okay, much better approach.  Mine is the prototype to show that it can work, yours can be the elegant solution that goes to prod.  😂

**Comment by RobSmyth** [comment]
Excellent. That sounds great. :-)

**Comment by RobSmyth** [comment]
Take a look at my branch. All I will be doing today. I also updated the event builder `WithEvent` method to take the event member name:

```
        device.HasEvent(config => config.OnModel(model)
                                        .WithEvent(nameof(QuickStartDemoModel.TestButtonPressed))
                                        .WithEventTypes(["press"])
                                        .WithFriendlyName("Test Event")
                                        .WithNodeId("test_event"));
```

OK?

Rob

**Comment by RobSmyth** [comment]
FYI ... added event to `QuickStartDemoApp.Program.cs` in the solution.

**Comment by AlmostInteractive** [comment]
Merged your changes into the feature branch.  Merged the resulting feature branch into (my) master.  Continued work on the feature branch.  I added a helper class, `HaEvent`, to replace the regular event.  It made things a little messy on the library side (see `EntityBuilderBase.cs` and `EntityDomainConfigMqttSourceFileGenerator.cs`) but provides a really nice helper class to use on the client.  Maybe you know how to make this cleaner?

I updated the QuickStartDemoApp with the changes.  In my local app, I have in the model class:

```
public HaEvent KeyboardEvent = new(["press"]);
...
device.HasEvent(config => config.OnModel(this)
            .WithEvent(nameof(KeyboardEvent))
            .WithFriendlyName("Test Event")
            .WithNodeId("test_event"));
```
and then raise the event with:
`_model.KeyboardEvent.Fire("press", new Dictionary<string, string> { { "param1", "value1" } });`

which works really nicely.  😃  The resulting Mqtt message triggered the HA event with the payload:
```
{
  "event_type": "press",
  "param1": "value1"
}
```

From the client usage side it feels so nice and clean!  Let me know what you think.  Once you think it's good enough, we can create a PR to get it into your repo.  👍 This is worth releasing a new NuGet version for!

**Comment by RobSmyth** [comment]
Nice. I think it has come along nicely.

But I'm not sure about `EventHandler` as:

* It exposes a public `Fire` method.
* Is designed to send multiple key-value pairs.

`Fire` being public means that any other code can fire that event.
I would want the model to have sole ownership of when it is fired.
An event does that. It is possible to subscribe to an event but only the owner can fire it.

As the library is intended for use within other apps I'm wondering if it may be better to use the conventional event approach as per C# guidelines (https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/events/how-to-publish-events-that-conform-to-net-framework-guidelines).
It makes the code easier to understand for anybody (new) looking at it and it enforces that only the owning class can fire the event.

Then, if it requires an argument I would implement an event like:
```
    public event EventHandler<HassEventArgs> MyEvent;
```

`EventHandler` nicely allows for it to pass key-value pairs.
The N2H intent is to separate HASS/MQTT from the client app and I see the event as being "pressed", "reset", etc events for a single entity.
Not sure how HASS would use the multiple values are to be used and how we send a MQTT configuration to allow HASS to parse these before any event is fired.
Perhaps these are attributes?
Can you give a use case.


Nice to hear your thinking.

**Comment by RobSmyth** [comment]
After typing the prior long post I realise it comes down to 2 questions to understand `HaEvent`:

- Do we want to expose a `Fire` method?
- Do we, and can we, send multiple key/value pairs (or use Attributes instead)?

These are what `HasEvent` is providing.

**Comment by RobSmyth** [comment]
Still thinking ... or are HASS events and buttons very different in that an event sends data. JSON?

**Comment by AlmostInteractive** [comment]
Hm, if we make the event not public (that's the only way to make it not publicly fireable yet still contained in a nice neat box) then we can't attach the listeners without passing in a pair of functions, like the prototype did.  We could require that they be named a certain way, like `{{HaEventMemberName}}Subscribe` and then search for those function by name, then they'd have to be implemented by the Model.  Dunno if that feels better or worse.  Really kind of a trade.

Events must have `event_type: string` and can optionally have any other data in the payload.  I don't know if everything in HA can parse and utilize the Event payload, I haven't tried, but I expect it should.  I'm pretty sure (and will shortly confirm) NodeRed will get it, though, which serves a lot of good for me.

> Still thinking ... or are HASS events and buttons very different in that an event sends data. JSON?

I don't know about Buttons for sure, but the docs don't mention any additional payload.  They do for Events: https://www.home-assistant.io/integrations/event.mqtt/

**Comment by RobSmyth** [comment]
BTW ... I'm catching up with MQTT events. I've not been playing with them. I was confused between buttons and events.

C# events look after that. The event is public and it cannot be fired outside of the class. The code I had was subscribing to the event without helper methods and was using a C# `event`.

Here is how to attach a listener to and event without helper methods:

```csharp
public class MyModel
{
    public event EventHandler MyEvent;

    public void SomeMethodThatFiresEvent()
    {
        var myEvent = MyEvent;
        if (myEvent != null)
        {
            myEvent(this, EventArgs.Empty); // we can fire the event
        }
    }
}

public class Client
{
    public Client(MyModel myModel)
    {
        myModel.MyEvent += MyModel_MyEvent; // SUBSCRIBING HERE
    }

    private void MyModel_MyEvent(object? sender, EventArgs e)
    {
        // do something
    }
}
```

As for the payload, for N2H to be a generic tool it needs to have a common parsing for payload and attributes. So maybe:

```json
{
   event:  
  {
      :
  }
   data:   []
}
```

`event` holds the event value json and `data` is a dictionary of attributes (as per other N2H entities).

Buttons (I think) are different, I know know, they define a specific name for the action.

> payload_press string (Optional, default: PRESS)
>
> The payload To send to trigger the button.

**Comment by AlmostInteractive** [comment]
In that code, `MyEvent` is public so anyone can reach in and Invoke it, right?  It doesn't seem like it improves encapsulation.  It's not much, but the HaEvent class provides setting event types without `.WithEventTypes` which is nice, and will also give you a happy warning (exception) if you try to fire an event type that won't be recognized and you're left figuring out why HA isn't seeing the event that you _know_ is being transmitted because _you can see it_ in MQTTX.....   I may have added this feature as a result of some personal frustration I had while testing....

The payload for Events must be at least
`{ "event_type": string }`
It can't be nested or omitted or it won't trigger.  We could do:

```
{
  "event_type": string,
  "data": {}
}
```
and still the key value pairs under data if you like.  Shouldn't have any negative effects on HA/NR functionality.  Just looking for values in a different spot.

**Comment by RobSmyth** [comment]
C# events support values. e.g: [`EventHandler<MyEventArgs>`](https://learn.microsoft.com/en-us/dotnet/api/system.eventhandler-1?view=net-8.0).

So you can do the same thing with the standard `event`. I would propose (maybe ... thinking) that we provide a base `MyEventArgs` that supports json serialisation. The developers can implement any programmatic arguments they like ... needs thinking.

Yea the payload structure looks consistent with those used for other entities.

> and still the key value pairs under data if you like.

It depends on what they are. That is are they attributes. `event_type` can hold a dictionary or whatever if it is to be programmatically handled in HASS.

What is your use case? Can you describe the data you want to send?

**Comment by AlmostInteractive** [comment]
From what I can tell without experimenting, `event_type` in HA can only be a string and `event_types` which it matches against says "A list of valid event_type strings."  Could be worth experimenting with though.

I plan to send various data with my events.  Duration, key values, event codes, text messages, so I'll be utilizing the dictionary nature of event payloads.  They'll be designed at run time by the user.

I guess we could make HaEvent take a `T` that defines the EventArgs and then make it abstract so they have to fill out the code to add args into the EventArgs object to send.  Or it could take a `data: string` and just shove pre-serialized JSON into it, leaving the reader to deserialize the JSON into its original form.  That's an option.

**Comment by RobSmyth** [comment]
Oh yes ... event_types is mandatory, for events.

Can you give more of a use cases like:

1 - Text message: has text and event code. Event code is an attribute? How would it be consumed in HASS?

2 - Press: Send a press event. Duration as an attribute?   <--- if so would this be better implemented as a button?

> They'll be designed at run time by the user.

This scares me a bit :-)  I see N2H about configuration as independent of the app as possible with the client code not needing to know about the MQTT stuff. 

Hmm. You're the hands on here ... I'm still thinking.

Use cases would help me heaps.

**Comment by AlmostInteractive** [comment]
I have `KeyboardEvent` which reacts when certain keyboard presses happen.  They're tied to user configurable event payloads.  Eg:

CTRL + Play = `{ "event_type": "press", "device": "media_player.chromecast", "cmd": "play"}`
CTRL + L = `{ "event_type": "press", "zone": "LivingRoom", "cmd": "toggle"}`
CTRL + M = `{ "event_type": "press", "script": "play_music"}`

I achieve these all with the `Dictionary<string, string>` as a general solution, but like I said, I can create that app-side and derive my own class from `HaEvent` which we can make super generic and abstract.

Events and Buttons differ in directionality.  Event:  App -> HA.  Button: HA -> App.  You can't trigger a button press via MQTT message (believe me, I tried).

**Comment by AlmostInteractive** [comment]
At the end of the day, it's your project!  Lets make it how you like it and I'll adapt my app to fit.  😄

**Comment by RobSmyth** [comment]
I see, ~~I think are 3 different events?~~

That helps improve my understanding. Thx.

> You can't trigger a button press via MQTT message

Wow ...that blows a lot of my thinking.


I need somebody hands on ... thank you. Big help. Nothing decided in my mind.

**Comment by RobSmyth** [comment]
Hows about:

```

CTRL + Play = { "event_type": "play", "value": "media_player.chromecast" }
CTRL + L = { "event_type": "toggle", "value": "LivingRoom" }
CTRL + M = { "event_type": "run_script", "value": "play_music"}
```

**Comment by AlmostInteractive** [comment]
Unless I dump JSON into "value", essentially turning it into a Dictionary, there'll be use cases that fall outside the `"value": string` format.  For example, one I haven't written yet, might look for something like this:

```
{ "event_type": "run_script", "value": "play_music", "search": "Red Hot", "search_mode": "Artist", "playlist_type": "random"}
{ "event_type": "run_script", "value": "play_music", "search": "Red Hot", "search_mode": "All"}
{ "event_type": "run_script", "value": "play_music", "playlist_type": "random"}
```
With any number of variable combinations requiring a new configuration and class.  I could, but I wouldn't; I'd implement a DictionaryEventArgs on the client side and use that.  😂 

I could dump it all into JSON and shove that in "value" but that seems less user friendly.  As long as I can get a key-value pair somewhere into the event to be parsed and sent along to HA, I'll be happy.

**Comment by RobSmyth** [comment]
My last post today, see ya.

Further brain storming...

```

public class HassEventArg
{
    public string Key { get; }

    public HassEventArg(string key)
    {
        Key = key;
    }
}

public class HassEventArgs : EventArgs
{
    public HassEventArgs(string eventType, params (string key, string jsonValue)[] args)
    {
        EventType = eventType;
        JsonKeyValuePairs =  args.ToDictionary(k => k.key, v => v.jsonValue);
    }

    /// <summary>
    /// Key value pairs that are essential event information. Does not include attributes.
    /// </summary>
    /// <remarks>
    /// Send in MQTT payload as:
    /// <code>
    /// {
    ///     event_type: &lt;EventType&gt;,
    ///     args: {
    ///         &lt;Key1&gt;: &lt;Value1&gt;,
    ///         &lt;Key2&gt;: &lt;Value2&gt;
    ///             :
    ///     },
    ///     data: {
    ///         -- Attributes key value pairs --
    ///     }
    /// }
    /// </code>
    /// </remarks>
    public string EventType { get; }

    /// <summary>
    /// The event payload. Excluding attributes which are handled separately.
    /// </summary>
    /// <remarks>
    /// Payload described in <see cref="EventType"/>.
    /// </remarks>
    public IDictionary<string, string> JsonKeyValuePairs { get; }
}

public class ExampleModel
{
    public event EventHandler<HassEventArgs> MyEvent;

    private void InternalMethod()
    {
        MyEvent(this, new HassEventArgs("run_script", 
                                        ("script_name", "play_music"), 
                                        ("search", "Red Hot"),          // Nice choice!!
                                        ("search_mode", "Artist"), 
                                        ("playlist_type", "random")));

        // -- OR - Same code, author's choice --

        MyEvent(this, new HassEventArgs("run_script",
                                        ("script_name", "play_music"),
                                        ("script_args", """
                                                        [
                                                           search: "Red Hot",
                                                           search_mode: "Artist",
                                                           playlist_type: "random",
                                                        ]
                                                        """)));
    }
}

```

**Reply by AlmostInteractive** [reply]
Looks fine.  Questions from a tired coder:

1. What benefit does this have over supplying a Dictionary?
2. Does it have to be a JSON string?  it seems like any string would work.  "[" isn't valid JSON I think it'll go over the wire just fine.
3. `MyEvent` is still public so it didn't solve for the encapsulation/model-only control that you were hoping for earlier, right?

**Reply by RobSmyth** [reply]
> Looks fine. Questions from a tired coder:
>
>    What benefit does this have over supplying a Dictionary?

Not a lot. I like that the user does not need to declare a dictionary.

>    Does it have to be a JSON string? it seems like any string would work. "[" isn't valid JSON I think it'll go over the wire just fine.

My intent was that the value could be any string or as JSON. The keys and values would be formatted as JSON anyway. JSON may be used to allow for common args parsing my HA via MQTT template member.

>    MyEvent is still public so it didn't solve for the encapsulation/model-only control that you were hoping for earlier, right?

Actually it does ... another class cannot invoke a class's public event. It is not a delegate or action. Try it, the code will not compile. An event can take whatever arguments you like. Even when an event is public it can only be fired by the owning class (the "publisher").  It solves encapsulation.

The configuration of valid event types moves to the N2H configuration code. Runtime can still assert a valid type.

**Reply by AlmostInteractive** [reply]
I did not know that.  Thanks!  I'll do a version with tuples instead of a dict and see how it feels.  

On the note of encapsulation, I made the `HaEvent` private in the model class and am still able to access it via reflection.  I just had to add a method on the model to allow the firing and now I think it's the best of both worlds.  😄

**Reply by AlmostInteractive** [reply]
I put a new version into my master branch that supports both tuples and dicts with a couple nice `Fire` overloads.  Dictionaries feel more comfortable to me (cuz I'm old) and tuples are more comfy for you (cuz you're cool) and now we can use both.  😃

**Comment by AlmostInteractive** [comment]
I merged that branch into my main.  We can do some refinements (such as the ones you proposed) before we PR it into your code.

I checked with HA to see if it could handle a kvp event payload and it worked, as expected.

```
hacommsclient/hacomms/test_event
{
  "event_type": "press",
  "param1": "value1"
}
```
triggered:
![image](https://github.com/user-attachments/assets/fabdd5bc-b27a-4d76-bf33-61699f05e964)

```
event_type: state_changed
data:
  entity_id: event.hacomms_test_event
  old_state:
    entity_id: event.hacomms_test_event
    state: "2024-07-27T17:26:11.245+00:00"
    attributes:
      event_types:
        - press
      event_type: press
      param1: value1
      friendly_name: HAcomms Test Event
    last_changed: "2024-07-27T17:26:11.245807+00:00"
    last_reported: "2024-07-27T17:26:11.245807+00:00"
    last_updated: "2024-07-27T17:26:11.245807+00:00"
    context:
      id: 01J3TKN11DSED6VYQPAJN9V7MR
      parent_id: null
      user_id: null
  new_state:
    entity_id: event.hacomms_test_event
    state: "2024-07-27T17:26:48.448+00:00"
    attributes:
      event_types:
        - press
      event_type: press
      param1: value1
      friendly_name: HAcomms Test Event
...
```

**Reply by RobSmyth** [reply]
Very nice & informative. So the event's state is the timestamp and the event type etc are passed as attributes. 

I wonder if this means that N2H attributes (`HasAttribute`) would not apply.

Much I do not know here.

**Reply by AlmostInteractive** [reply]
I dunno!  :-D  We're both figuring this stuff out together.

**Comment by AlmostInteractive** [comment]
So I've got this version built into [my app](https://github.com/AlmostInteractive/HAcomms) and I'm loving it.  I hooked up the keyboard `CTRL + MediaPlayPause` to fire an HaEvent, handled it in an automation, and wired it up to my home speakers.  😄 

![image](https://github.com/user-attachments/assets/d4188907-9bfb-400e-bcae-4d0da6846315)

Super happy with this!

**Comment by RobSmyth** [comment]
I wonder if you have used other parts of N2H. I want to get an idea of what bits have been used by somebody other than me.

**Comment by AlmostInteractive** [comment]
This is my first time using your library and I'm glad I could contribute something to it.  Here's my usage, if you're curious:

https://github.com/AlmostInteractive/HAcomms/blob/master/HAcomms/Models/HAcommsModel.cs

**Comment by AlmostInteractive** [comment]
Hey @RobSmyth  sorry I've been incommunicado.  I just moved and was without power.  It's a whole thing.

Anyway!  What do you think about my current branch?  It's been working great for me, except for one thing that isn't new to my code and I don't know yet how to get around.  Events and Buttons get triggered when the application starts because they get updated.  This results in a phantom activation of my Events and also on the Button handler app-side.   I have an idea about making a "HaPulseEvent" which will set state and then set it back to "unknown" moments later which I think will help my use case.  But anyway, do you feel it's ready to merge into your code base?

**Comment by RobSmyth** [comment]
Hi!

I've been working on automated versioning for N2H, and other projects, so I've not got back to looking at this. But yes, I do want to merge it into N2H. I'm not sure about how to do the event bit, so I propose you raise a PR to a feature branch off master and I can make what changes I like before merging to master. Does that sound good?

Having a look at the Events and Buttons triggering on start.

**Comment by AlmostInteractive** [comment]
Neat.  👍   I opened the PR.  If you figure out the trigger-on-start thing, that would be cool, but I think it's an HA thing.  I tested with my S3 Box3 running ESPHome and its buttons triggered at startup too.  Doesn't mean there isn't a hack-around though.  😄

**Comment by RobSmyth** [comment]
On the triggering on start ... not sure I have the code, can you point me at some lines of how/why they get updated at start? I see `FireKeyboardComboEvent` in your code fragment but not how it is called.

**Reply by AlmostInteractive** [reply]
All entries are State based (even buttons) so when the app starts up, it sends the birth message to set the current state.  When the app closes, the lwt message sets the state (correctly) to `unavailable`:

![image](https://github.com/user-attachments/assets/4f9d9c5a-be75-4aef-a87b-80bd72fb188e)

When the app starts up, the birth message sets those as available which updates their state to whatever (eg: not `unavailable`) which for buttons and events triggers an actionable update.  

Steps to repro:

1. Create a button entity with a command method
2. Set a break point in the command method
3. Debug run your app
4. See that when it connects to HA, the command entity gets called immediately as if the button were pressed

(Same thing for events, but the triggering happens on the HA side.)

**Reply by RobSmyth** [reply]
Ah yes ... I think all entities are currently "retained" so the event state will show as retained on the broker. I think an event should not be retained.

I think most entity MQTT config messages have a 'retain' property. Hmmm.

**Reply by AlmostInteractive** [reply]
It's not that.  Events (and buttons?) ignore the "Retained" flag if set.

**Reply by RobSmyth** [reply]
After looking at the code I see that a retain option is used on the client connection. An MQTT switch, device trigger, & button do have a retain property but event does not (makes sense). So my guess is that if the connection did not use the retain option would the event may work as you expect. 

If so, the retain option will need to be moved to the entities that need it (switch etc).

**Reply by RobSmyth** [reply]
> It's not that. Events (and buttons?) ignore the "Retained" flag if set.

Your running actual tests, but I note from https://www.home-assistant.io/integrations/event.mqtt/

> MQTT events are stateless. For example, a doorbell does not have a state like being “on” or “off” but instead is momentarily pressed.

An the MQTT event config message does not have a retain property. But if the client configuration overrides that on the broker ... dunno. A test event on my broker is showing as retained ... dunno what version of the software that was though.

![image](https://github.com/user-attachments/assets/a7ff6921-fddd-4816-9ab5-480f9fd6bc42)

**Reply by RobSmyth** [reply]
Dunno ... might be worth a quick test ...

![image](https://github.com/user-attachments/assets/89122a77-ab3f-4e40-b8e1-52f4ee22de94)

**Reply by RobSmyth** [reply]
I just did a very quick test of removing that retain option. It did remove the retained flag from the event on the broker after I fired another event.

So that option is making the broker retain event state even though HASS does not expect it to.

**Reply by AlmostInteractive** [reply]
Strange, in MQTTX I don't see the retained flag set on any of my Event fires.

**Reply by RobSmyth** [reply]
I'm so confused ... I've found MQTTX has some strange behavior. Often it does not show the retain flag until you restart it :-)

Also ... the flag usually disappears on update. I'm no longer sure what is happening.

**Reply by RobSmyth** [reply]
I think I've not caught up to what you have found and were trying to tell me LOL 😄 

Looking at HASS logs I do not see any event message on start=up. Only the "online" message. HASS is displaying the last event as a state .... aaagggh!

I found a couple post were people report that events are fired on HASS start-up and people have done workarounds.

Perhaps add a "None" event value and send that immediately after the real event. Ouch.

**Reply by AlmostInteractive** [reply]
I actually wrote exactly that, a "pulse" option on the even that would send an "unknown" state 100ms after the event.  It helps with NodeRed flows because you can filter those out, but it doesn't help with HASS (without filtering) since it still counts as a state change.

https://github.com/AlmostInteractive/Net2HassMqtt/tree/pulseEvent

**Comment by RobSmyth** [comment]
Great! Merged, I'll sort out my versioning and get another early dev release out soon.

**Comment by RobSmyth** [comment]
Even if not related to your event on start issue ... I'm fixing the retain flag now. Moving from client to each event that supports retain.

**Reply by RobSmyth** [reply]
For what little it is worth I've attached my HASS logs that show real events "A" / "B" and then a shutdown and restart.

[home-assistant_mqtt_2024-08-03T07-01-51.239Z.log](https://github.com/user-attachments/files/16478462/home-assistant_mqtt_2024-08-03T07-01-51.239Z.log)

**Comment by RobSmyth** [comment]
Hi @AlmostInteractive,

I've updated the code the N2H [button branch](https://github.com/NoeticTools/Net2HassMqtt/tree/Adding-button-entity). If you would like to have a look the class `QuickStartDemoModel` gives you a quick idea of the direction I'm going.

Interesting ... I now understand that HASS is looking for the `event_type` property and adding all others as attributes. This results in 2 N2H attribute sources:

1. Attribute properties that N2H always reads when the event is being published.
2. Attributes that the user's application (client) code adds to the event. I now understand that these are event type specific.

These events are now merged into the message and sent the same as other messages.

I'm now making the event configurable as a pulse as a HASS workaround. WIP. Thinking that it will treat the first event type as the steady (or idle) state and always publish immediately after an event.  I feel dirty doing a work around, but it is a configuration option :-).

**Reply by AlmostInteractive** [reply]
[HassDeviceClassesSourceGenerator/FileGenerators/EntityDomainConfigMqttSourceFileGenerator.cs](https://github.com/NoeticTools/Net2HassMqtt/compare/main...Adding-button-entity#diff-f0d0c396c236734f38fd50e24b6a6fc12998d55e2e6d01727ddab195e1636eba) line 114 looks to have a typo:  `-` instead of `~`.

I like the `HassEventArgsFactory` 😄

**Reply by RobSmyth** [reply]
> [HassDeviceClassesSourceGenerator/FileGenerators/EntityDomainConfigMqttSourceFileGenerator.cs](https://github.com/NoeticTools/Net2HassMqtt/compare/main...Adding-button-entity#diff-f0d0c396c236734f38fd50e24b6a6fc12998d55e2e6d01727ddab195e1636eba) line 114 looks to have a typo: `-` instead of `~`.
> 
> I like the `HassEventArgsFactory` 😄

The `-` is about line spacing. On consumes white space on button not line breaks, the other consumes lines.

**Reply by RobSmyth** [reply]
A little more work done:

Remove need to configure event types individually (`WithEventTypes<-enum->(`)):

```
        device.HasEvent(config => config.OnModel(model)
                                        .WithEvent(nameof(QuickStartDemoModel.Event3))
                                        .WithEventTypes<Event3Types>()   ...
```

Now supports optional clear after each event (`WithFirstEventTypeSentAfterEachEvent`):

```
        device.HasEvent(config => config.OnModel(model)
                                        .WithEvent(nameof(QuickStartDemoModel.Event2))
                                        .WithEventTypes(["Clear", "PressedC", "PressedD"])
                                        .WithFirstEventTypeSentAfterEachEvent()  ...
```

Let me know if this does not look right to you.

Now I need to test it  :-) ... handle events is Home Assistant :-)
